README.md
znc cookbook
============

Installs and configures ZNC, an IRC bouncer.

Usage
=====

At least one user needs to be defined in either

    - the `node[:znc][:users]` attribute
    - or in a `users` data_bag

Including `recipe[znc::default]` in your run-list will install ZNC
via the target OS's package manager.

Requirements
============

## Chef

Tested on Chef 11 but newer and older version should work just fine. File an
[issue][issues] if this isn't the case.

## Platform

The following platforms have been tested with this cookbook, meaning that the
recipes and LWRPs run on these platforms without error:

* Ubuntu 10.04, 12.04, 14.04

CentOS 6 mostly works, as ZNC gets installed and configured, but as of this
writing, the service does not automatically start at the end of the
`chef-client` run.

# Cookbooks

This cookbook depends on the following external cookbooks:

* [build-essential](http://community.opscode.com/cookbooks/build-essential) (Opscode)
* [user](https://github.com/fnichol/chef-user)

Recipes
=======

## default

Include the default recipe in a run list, to get `znc`.  By default `znc` is installed from packages but this can be changed by using the `install_method` attribute.

## package

This recipe installs ZNC from packages.

## source

This recipe installs ZNC from source.

Attributes
==========

System Attributes
-----------------

Key                         | Type   | Description                            | Default
--------------------------- | ------ | -------------------------------------- | --------------------
`['znc']['install_method']` | String | `package` or `source`                  | `package`
`['znc']['user']`           | String | user used to install and run ZNC       | `znc`
`['znc']['group']`          | String | group used to install and run ZNC      | `znc`
`['znc']['functions']`      | String | path to lsb/init.d functions           | _platform dependent_

ZNC Attributes
--------------

Key                          | Type          | Description                        | Default
---------------------------- | ------------- | ---------------------------------- | -------------------------------------
`['znc']['data_dir']`        | String        | absolute path to store ZNC data    | _platform dependent_
`['znc']['conf_dir']`        | String        | absolute path to store ZNC configs | `#{znc['data_dir']}/configs`
`['znc']['log_dir']`         | String        | absolute path to store ZNC logs    | `#{znc['data_dir']}/moddata/adminlog`
`['znc']['module_dir']`      | String        | absolute path to store ZNC modules | `#{znc['data_dir']}/modules
`['znc']['users_dir']`       | String        | absolute path to store ZNC users   | `#{znc['data_dir']}/users
`['znc']['users']`           | String Array  | list of users to add to ZNC        | `nil` (`[]` when using chef-solo)
`['znc']['port']`            | String        | port for ZNC to listen on          | `+7777`
`['znc']['skin']`            | String        | skin to use with ZNC               | `dark-clouds`
`['znc']['max_buffer_size']` | Integer       | maximum buffer size                | 500
`['znc']['modules']`         | String Array  | ZNC modules to load                | ['webadmin', 'adminlog']

Testing
=======

This cookbook ships with full support for the new alpha version of Opscode's
`test-kitchen`. [Fletcher Nichol's](https://github.com/fnichol) excellent Jamie
integration test runner was
[recently merged into the test-kitchen codebase](http://lists.opscode.com/sympa/arc/chef-dev/2013-01/msg00038.html).
This merge and all new feature work are now taking place in the
[1.0 branch of test-kitchen](https://github.com/opscode/test-kitchen/tree/1.0).
Even though many community members have been dogfooding the new `test-kitchen`
codebase and it has proven quite stable, it should still be regarded as
pre-release code and YMMV.

Running integration with test-kitchen is easy. First we'll assume you have a
sane cookbook development toolchain installed which includes:

* Git
* Ruby 1.9.x
* Bundler
* VirtualBox 4.x

If you need help setting up this toolchain, take a read through the
"[System Setup](http://vialstudios.com/guide-authoring-cookbooks.html#system_setup)"
section of [Jamie Winsor's](https://github.com/reset) excellent cookbook
authoring guide.

First install all gem dependencies with Bundler:

```shell
$ bundle install --binstubs
```

Bundler will install all of the dependent RubyGems and guarantee that you have
the right versions. Now it's time to get your test on:

```shell
$ bundle exec kitchen test
```

This command will do the following across every platform/version this cookbook
supports:

* Use Vagrant to provision a platform-specific VM.
* Add a `recipe[znc]` entry to the instance's run_list.
* Converge the VM with `chef-solo`.

By default, any VM that successfully converges will automaticallly be cleaned
up. The full set of `kitchen` subcommands can be viewed by running:

```shell
$ bundle exec kitchen help
```

Development
===========

* Source hosted at [GitHub][repo]
* Report issues/Questions/Feature requests on [GitHub Issues][issues]

I <3 pull requests! Make sure your patches are well tested. Ideally create a
topic branch for every separate feature/fix you make.

License and Author
==================

Author:: [Seth Chisamore][schisamo] (<schisamo@gmail.com>) [![endorse](http://api.coderwall.com/schisamo/endorsecount.png)](http://coderwall.com/schisamo)

Copyright (c) 2011-2013, Seth Chisamore

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

[schisamo]:      https://github.com/schisamo
[repo]:          https://github.com/schisamo-cookbooks/znc
[issues]:        https://github.com/schisamo-cookbooks/znc/issues


CHANGELOG.md
# 0.2.1 (June 15, 2014)

Added unit & integration tests with ChefSpec and Serverspec.

# 0.2.0 (June 18, 2013)

Incorporated [fnichol](https://github.com/fnichol)'s pull-requests:

    - [Support chef-solo and/or reading users from an attribute.](https://github.com/schisamo-cookbooks/znc/pull/5)
    - [Only signal znc process if process is running.](https://github.com/schisamo-cookbooks/znc/pull/6)
    - [Wrap user hash user['znc']['modules'] in Array().](https://github.com/schisamo-cookbooks/znc/pull/7)

# 0.1.0 (March 2, 2013)

Pulled from [schiasmo-cookbooks/znc](https://github.com/schisamo-cookbooks/znc):

    - [Initial commit, rewrite coming very soon!](https://github.com/schisamo-cookbooks/znccommit/7e9c14e1fca9fad464b7011aac5d57d00caca68e)


test/unit/spec/source_spec.rb
# Encoding: utf-8
require_relative 'spec_helper'

describe 'znc::source' do
  PLATFORMS.each do |platform, versions|
    versions.each do |version|
      context "on #{platform} #{version}" do
        let(:chef_run) do
          ChefSpec::Runner.new(platform: platform, version: version)
            .converge(described_recipe)
        end

        let(:checksum)     { '6dafcf12b15fdb95eac5b427c8507c1095e904b4' }
        let(:source_dest)  { "#{Chef::Config[:file_cache_path]}/znc-1.4.tar.gz" }

        before { stub_command('which znc').and_return(nil) }

        it 'installs the build-essential package' do
          expect(chef_run).to include_recipe('build-essential')
        end

        packages =
          case platform
          when 'debian', 'ubuntu'
            %w( libssl-dev libperl-dev pkg-config libc-ares-dev )
          else
            %w( openssl-devel perl-devel pkgconfig c-ares-devel )
          end

        context 'installs' do
          packages.each do |package|
            it package do
              expect(chef_run).to install_package(package)
            end
          end
        end

        context 'ZNC is not installed' do
          it 'downloads the source tarball' do
            expect(chef_run).to create_remote_file(source_dest).with(
              checksum: checksum,
              mode: '0644'
            )
          end

          it 'builds ZNC' do
            expect(chef_run).to run_bash('build znc')
              .with_cwd(Chef::Config[:file_cache_path])
          end
        end

        context 'ZNC is installed' do
          before { stub_command('which znc').and_return('/usr/bin/znc') }

          it 'does not download the source tarball' do
            expect(chef_run).to_not create_remote_file(source_dest)
          end

          it 'does not build ZNC' do
            expect(chef_run).to_not run_bash('build znc')
          end
        end
      end
    end
  end
end


test/unit/spec/spec_helper.rb
# Encoding: utf-8
require 'chefspec'
require 'chefspec/berkshelf'

::LOG_LEVEL = :fatal
::UBUNTU_OPTS = {
  platform: 'ubuntu',
  version: '14.04',
  log_level: ::LOG_LEVEL
}
::CHEFSPEC_OPTS = {
  log_level: ::LOG_LEVEL
}

PLATFORMS = {
  'ubuntu' => %w( 10.04 12.04 14.04 ),
  'centos' => %w( 6.4 6.5 )
}

at_exit { ChefSpec::Coverage.report! }


test/unit/spec/default_spec.rb
# Encoding: utf-8
require_relative 'spec_helper'

describe 'znc::default' do
  PLATFORMS.each do |platform, versions|
    versions.each do |version|
      context "on #{platform} #{version}" do
        %w( package source ).each do |install_method|
          context "installed via #{install_method}" do
            let(:chef_run) do
              ChefSpec::Runner.new(platform: platform,
                                   version: version) do |node|
                node.set['znc']['install_method'] = install_method
              end.converge(described_recipe)
            end

            let(:user)  { 'znc' }
            let(:group) { 'znc' }

            dir = Hash.new
            dir[:data]   = '/etc/znc'
            dir[:conf]   = "#{dir[:data]}/configs"
            dir[:module] = "#{dir[:data]}/modules"
            dir[:users]  = "#{dir[:data]}/users"

            let(:znc_conf_path) { "#{dir[:conf]}/znc.conf" }

            before do
              stub_command('pgrep znc').and_return('12345')
              stub_command('which znc')
                .and_return(nil) if install_method == 'source'
            end

            it "includes the `znc::#{install_method}` recipe" do
              expect(chef_run).to include_recipe("znc::#{install_method}")
            end

            it 'creates a znc user' do
              expect(chef_run).to create_user(user)
            end

            it 'creates a znc group' do
              expect(chef_run).to create_group(group)
            end

            context 'creates directories with znc user/group' do
              dir.each do |type, path|
                it type do
                  expect(chef_run).to create_directory(path).with(
                    owner: user,
                    group: group
                  )
                end
              end
            end

            it 'generates the pem file' do
              expect(chef_run).to run_bash('generate-pem').with(
                cwd: dir[:data],
                user: user,
                group: group,
                creates: "#{dir[:data]}/znc.pem"
              )
            end

            it 'generates an init script' do
              expect(chef_run).to create_template('/etc/init.d/znc').with(
                source: 'znc.init.erb',
                owner: 'root',
                group: 'root',
                mode: '0755'
              )
            end

            context 'ZNC is running' do
              it 'force save the config' do
                expect(chef_run).to run_execute('force-save-znc-config')
              end
            end

            context 'ZNC is not running' do
              before { stub_command('pgrep znc').and_return(nil) }

              it 'does not force save the config' do
                expect(chef_run).to_not run_execute('force-save-znc-config')
              end
            end

            it 'prepares to reload the config' do
              expect(chef_run.execute('reload-znc-config')).to do_nothing
            end

            it 'generates the config' do
              expect(chef_run).to create_template(znc_conf_path).with(
                source: 'znc.conf.erb',
                mode: '0666',
                owner: user,
                group: group,
                variables: { users: [] }
              )
            end

            it 'reloads the config' do
              expect(chef_run.template(znc_conf_path))
                .to notify('execute[reload-znc-config]').to(:run).immediately
            end

            context 'starts ZNC' do
              it 'enables' do
                expect(chef_run).to enable_service('znc')
              end

              it 'starts' do
                expect(chef_run).to start_service('znc')
              end
            end
          end
        end
      end
    end
  end
end


recipes/package.rb
#
# Cookbook Name:: znc
# Recipe:: package
#
# Copyright (c) 2011-2013, Seth Chisamore
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

case node['platform']
when 'macosx'
  # TODO: install via homebrew
else
  znc_pkgs = value_for_platform(
    %w( debian ubuntu ) => {
      'default' => %w( znc znc-dev znc-extra znc-webadmin )
    },
    'default' => %w( znc znc-devel )
  )

  if node['platform'] == 'ubuntu'
    os_version = node['platform_version'].to_f

    znc_pkgs.delete('znc-extra')    if os_version >= 14.04
    znc_pkgs.delete('znc-webadmin') if os_version > 10.04
  end

  znc_pkgs.each do |pkg|
    package pkg do
      action :install
    end
  end

end


attributes/source.rb
#
# Cookbook Name:: znc
# Attribute:: source
#
# Copyright (c) 2011-2013, Seth Chisamore
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

default['znc']['url'] = 'http://znc.in/releases'
default['znc']['version'] = '1.4'
default['znc']['checksum'] = '6dafcf12b15fdb95eac5b427c8507c1095e904b4'
default['znc']['configure_options'] = %W( --enable-extra )


recipes/module_colloquy.rb
#
# Cookbook Name:: znc
# Recipe:: module_colloquy
#
# Copyright (c) 2011-2013, Seth Chisamore
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# znc > 0.0.9 required...this means compiling from source on most platforms

remote_file "#{Chef::Config[:file_cache_path]}/colloquy.cpp" do
  source 'https://github.com/wired/colloquypush/raw/master/znc/colloquy.cpp'
  mode '0644'
  not_if { ::File.exist?("#{node['znc']['module_dir']}/colloquy.so") }
end

bash 'build colloquy znc module' do
  cwd Chef::Config[:file_cache_path]
  code <<-EOF
  znc-buildmod colloquy.cpp
  mv colloquy.so #{node['znc']['module_dir']}/
  EOF
  creates "#{node['znc']['module_dir']}/colloquy.so"
  notifies :run, 'execute[reload-znc-config]', :immediately
end


recipes/default.rb
#
# Cookbook Name:: znc
# Recipe:: default
#
# Copyright (c) 2011-2013, Seth Chisamore
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe "znc::#{node['znc']['install_method']}"

user node['znc']['user']
group node['znc']['group']

[node['znc']['data_dir'],
 node['znc']['conf_dir'],
 node['znc']['module_dir'],
 node['znc']['users_dir']
].each do |dir|
  directory dir do
    owner node['znc']['user']
    group node['znc']['group']
  end
end

bash 'generate-pem' do
  cwd node['znc']['data_dir']
  code <<-EOH
  umask 077
  openssl genrsa 2048 > znc.key
  openssl req -subj /C=US/ST=Several/L=Locality/O=Example/OU=Operations/CN=#{node['fqdn']}/emailAddress=znc@#{node['fqdn']} \
   -new -x509 -nodes -sha1 -days 3650 -key znc.key > znc.crt
  cat znc.key znc.crt > znc.pem
  EOH
  user node['znc']['user']
  group node['znc']['group']
  creates "#{node['znc']['data_dir']}/znc.pem"
end

template '/etc/init.d/znc' do
  source 'znc.init.erb'
  owner 'root'
  group 'root'
  mode '0755'
end

# ~FC003 alternative support for chef-solo
if Chef::Config[:solo] &&
    !run_context.cookbook_collection.include?('chef-solo-search')
  users = node['znc']['users']
else
  users = node['znc']['users'] || search(:users, 'groups:znc')
end

# znc doesn't like to be automated...this prevents a race condition
# http://wiki.znc.in/Configuration#Editing_config
execute 'force-save-znc-config' do
  command 'pkill -SIGUSR1 znc'
  action :run
  only_if 'pgrep znc'
end
execute 'reload-znc-config' do
  command 'pkill -SIGHUP znc'
  action :nothing
  only_if 'pgrep znc'
end

# render znc.conf
template "#{node['znc']['data_dir']}/configs/znc.conf" do
  source 'znc.conf.erb'
  mode '0666'
  owner node['znc']['user']
  group node['znc']['group']
  variables(
    users: users
  )
  notifies :run, 'execute[reload-znc-config]', :immediately
end

service 'znc' do
  supports restart: true
  action [:enable, :start]
end
